Firestore/core/src/firebase/firestore/local/leveldb_remote_document_cache.cc

```diff
@@ -48,12 +51,52 @@ using model::SnapshotVersion;
 using nanopb::ByteString;
 using nanopb::Message;
 using nanopb::StringReader;
+using util::BackgroundQueue;
+using util::Executor;
+
+/**
+ * An accumulator for results produced asynchronously. This accumulates
+ * values in a vector to avoid contention caused by accumulating into more
+ * complex structures like immutable::SortedMap.
+ */
+template <typename T>
+class AsyncResults {
+ public:
+  void Insert(T&& value) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    values_.push_back(std::move(value));
+  }
+
+  void Insert(const T& value) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    values_.push_back(value);
+  }
+
+  const std::vector<T>& Result() {
```

Comment:
I think that returning a reference to (mutex-) protected data is generally something to avoid. Even with the current very limited interface it could result in potentially misleading code:
```cpp
if (foo.Result().empty()) {
  // Is `foo` actually empty at this point? Not necessarily
  foo.Insert("abc");
}
```
It could become dangerous if the interface were ever to grow (e.g., the classic `if (!foo.empty()) foo.pop();` problem).

One way to achieve that would be to move conversion to the expected end result (`DocumentMap`, etc.) into this class. Given that this class, at this point at least, is very much tailored to the needs of this file and isn't even exposed outside of this translation unit, I don't think it's inappropriate.

Alternatively: how much is the current code buying by doing some insertions on the current thread (the empty ones) and some on the background thread? I'm pretty sure that if all insertions were to be on the background thread, the mutex (and this class by extension) would be unnecessary -- `WaitForAll` would create the necessary "happens-before" relationship (of course, TSan can check these assumptions). I'm not sure if that would be the case with the current code structure, though (with some insertions happening on the current thread).

Reply:
The issue is that it's background _threads_. The background executor is created with as many threads as are returned from `std::thread::hardware_concurrency`. Recent iPhones have 6 cores. My MacBook Pro has 12 (after hyperthreading). The mutex is required for contention among those threads, not just with the main worker.

As to moving the result-building into this class, that doesn't work well. In one case, we're building an `OptionalMaybeDocumentMap` based on `std::pair<DocumentKey, absl::optional<MaybeDocument>>`. In the other case we're building a `DocumentMap` based on `Document`. There's no unified way to process these. Meanwhile, a for-each style interface is cumbersome with the way our immutable maps accumulate.

The intent here is that all `Insert`s happen before the call to `Result` and that `Result` is only called once, after `BackgroundQueue::AwaitAll` has returned. There's nothing here constraining that to be the case but it's hard to ensure the cooperation of these classes because they're not aware of each other.

I could add a state machine here that prevents calling `Insert` after `Result` has been called but that feels like a hack. Short of that I'm having a hard time coming up with anything that substantively addresses the concern that doesn't add a lot of boilerplate in the caller.

Reply:
Ah, good point about multiple threads.

> The intent here is that <...> `Result` is only called once, after `BackgroundQueue::AwaitAll` has returned.

How about returning by value, then?

```cpp
std::vector<T> ConsumeResult() {
  std::lock_guard lock{m_};
  auto result = std::move(result_);
  result_.clear(); // Might be overkill
  return result;
}
```

It solves any potential concurrency problems and as an added benefit reflects the intended usage in the API itself. I don't expect the extra move to matter performance-wise (not to mention that accessing a variable using indirection also in theory could have some extra cost). What do you think?

Reply:
Oh, and I think it would enable actually moving the docs into the resulting map.

Reply:
I made it consume by value.

`immutable::SortedMap` cannot move into key or value. The complexity of making that work was stratospheric and I didn't bother...

Both DocumentKey and Document are reference-count bumps so I'm happy to leave this as-is.